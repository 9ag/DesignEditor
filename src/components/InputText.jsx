import React from 'react'
import classnames from 'classnames'
const mobxReact = require('mobx-react')

class InputText extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      showList: false
    }
  }

  queryChange = (e) => {
    this.props.queryChange(e.target.value)
  }

  choose = (data) => {
    this.props.choose(this.props.data, data)
    this.hideList()
  }

  showList = () => {
    this.setState({showList: true})
  }

  hideList = () => {
    this.setState({showList: false})
  }

  inputKeyDown = (e) => {
    // backspace
    if (e.keyCode === 8) {
      this.props.clear()
    }
  }

  render() {
    const {query, data, clear, type, id} = this.props
    const ulStyle = {display: this.state.showList ? 'block' : 'none'}
    const selected = data.find(item => item._selected)
    const inputShowText = selected ? selected.Name : query
    return (
      <div className="" onMouseLeave={this.hideList}>
        <div className="search-group">
          <input onMouseEnter={this.showList} onFocus={this.showList} value={inputShowText} onKeyDown={this.inputKeyDown} id={id}
             onChange={this.queryChange} name="TestReactiveWoNumber" type="text" className="form-control input-sm"
              placeholder={type} data-bind="value: woNum" data-autocomplete="ReactiveWoNumber" />
          <a onClick={clear} className="close"><i className="fa fa-close f12"></i></a>
        </div>
        <ul style={ulStyle} className="ui-autocomplete ui-front ui-menu ui-widget ui-widget-content ui-corner-all react-ipt-ul" id="ui-id-2">
          {
            data.map(item => {
              const cn = classnames({'input-sigle-active': item._selected, 'ui-menu-item': true})
              return <li key={item.Id} onClick={this.choose.bind(this, item)} className={cn} role="presentation">
                <a id="ui-id-32" className="ui-corner-all" >{item.Name}</a>
              </li>
            })
          }
        </ul>
      </div>
    )
  }
}

export default mobxReact.observer(InputText)