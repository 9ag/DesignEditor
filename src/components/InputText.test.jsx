
import { shallow, mount, render } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import * as enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
enzyme.configure({ adapter: new Adapter() });

import InputText from './InputText';
describe("InputText component test", () => {

  let choose = null;
  let queryChange = null;
  let clear = null;
  const data = [{
    Id: 1,
    Name: 'changsha',
    _selected: true,
  }, {
    Id: 2,
    Name: 'xiangtan',
  }, {
    Id: 3,
    Name: 'loudi',
  }];
  beforeEach(() => {
    choose = jest.fn();
    queryChange = jest.fn();
    clear = jest.fn();
  });

  it("show/hide dropdown list test", () => {
    const componentNode = renderer.create(<InputText data={data} choose={choose} clear={clear} queryChange={queryChange} />);
    let component = componentNode.toJSON();
    //  component.props.onMouseLeave();
    componentNode.root.children[0].props.onMouseLeave();
    expect(componentNode.root.instance.state.showList).toBe(false);
    component = componentNode.toJSON();
    expect(component).toMatchSnapshot();
    
    const input = componentNode.root.findByType('input');
    // const input = component.children[0].children[0];
    input.props.onMouseEnter();
    expect(componentNode.root.instance.state.showList).toBe(true);
    component = componentNode.toJSON();
    expect(component).toMatchSnapshot();
  });

  it("input querychange test", () => {
    const component = shallow(<InputText data={data} choose={choose} clear={clear} queryChange={queryChange} />);
    const event = {
      preventDefault() {},
      target: { value: 'search key' }
    };
    component.find('input').simulate('change', event);
    expect(queryChange).toBeCalledWith('search key');
  });

  it("dropdown list choose test", () => {
    const component = shallow(<InputText data={data} choose={choose} clear={clear} queryChange={queryChange} />);
    
    component.find('li').at(0).simulate('click');
    expect(choose).toBeCalledWith(data, data[0]);
  });

  it("input keydown test", () => {
    const component = shallow(<InputText data={data} choose={choose} clear={clear} queryChange={queryChange} />);
    const event = {
      preventDefault() {},
    };
    component.find('input').simulate('keydown', event);
    expect(clear.mock.calls.length).toBe(0);
    event.keyCode = 8;
    component.find('input').simulate('keydown', event);
    expect(clear.mock.calls.length).toBe(1);
  });

 
});