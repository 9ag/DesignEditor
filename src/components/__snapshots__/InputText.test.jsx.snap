// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputText component test show/hide dropdown list test 1`] = `
<div
  className=""
  onMouseLeave={[Function]}
>
  <div
    className="search-group"
  >
    <input
      className="form-control input-sm"
      data-autocomplete="ReactiveWoNumber"
      data-bind="value: woNum"
      name="TestReactiveWoNumber"
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseEnter={[Function]}
      type="text"
      value="changsha"
    />
    <a
      className="close"
      onClick={[MockFunction]}
    >
      <i
        className="fa fa-close f12"
      />
    </a>
  </div>
  <ul
    className="ui-autocomplete ui-front ui-menu ui-widget ui-widget-content ui-corner-all react-ipt-ul"
    id="ui-id-2"
    style={
      Object {
        "display": "none",
      }
    }
  >
    <li
      className="input-sigle-active ui-menu-item"
      onClick={[Function]}
      role="presentation"
    >
      <a
        className="ui-corner-all"
        id="ui-id-32"
      >
        changsha
      </a>
    </li>
    <li
      className="ui-menu-item"
      onClick={[Function]}
      role="presentation"
    >
      <a
        className="ui-corner-all"
        id="ui-id-32"
      >
        xiangtan
      </a>
    </li>
    <li
      className="ui-menu-item"
      onClick={[Function]}
      role="presentation"
    >
      <a
        className="ui-corner-all"
        id="ui-id-32"
      >
        loudi
      </a>
    </li>
  </ul>
</div>
`;

exports[`InputText component test show/hide dropdown list test 2`] = `
<div
  className=""
  onMouseLeave={[Function]}
>
  <div
    className="search-group"
  >
    <input
      className="form-control input-sm"
      data-autocomplete="ReactiveWoNumber"
      data-bind="value: woNum"
      name="TestReactiveWoNumber"
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseEnter={[Function]}
      type="text"
      value="changsha"
    />
    <a
      className="close"
      onClick={[MockFunction]}
    >
      <i
        className="fa fa-close f12"
      />
    </a>
  </div>
  <ul
    className="ui-autocomplete ui-front ui-menu ui-widget ui-widget-content ui-corner-all react-ipt-ul"
    id="ui-id-2"
    style={
      Object {
        "display": "block",
      }
    }
  >
    <li
      className="input-sigle-active ui-menu-item"
      onClick={[Function]}
      role="presentation"
    >
      <a
        className="ui-corner-all"
        id="ui-id-32"
      >
        changsha
      </a>
    </li>
    <li
      className="ui-menu-item"
      onClick={[Function]}
      role="presentation"
    >
      <a
        className="ui-corner-all"
        id="ui-id-32"
      >
        xiangtan
      </a>
    </li>
    <li
      className="ui-menu-item"
      onClick={[Function]}
      role="presentation"
    >
      <a
        className="ui-corner-all"
        id="ui-id-32"
      >
        loudi
      </a>
    </li>
  </ul>
</div>
`;
